pr

jexxor opened the pull request.

jexxor committed:
init filter_test

jexxor committed:
package refactor, added save file

jexxor committed:
refactor filter main functions

jexxor committed:
update filter Insert and Exists

jexxor committed:
Update filters, bitset. Processing interface

jexxor committed:
feat: primitive cli

jexxor committed:
implemented hashes

jexxor committed:
update readme

jexxor committed:
removed author hashes

jexxor committed:
code refactor, negative index bugfix

jexxor committed:
cli refactor

jexxor committed:
Update README.md

jexxor committed:
Update README.md

jexxor committed:
Rm all

jexxor committed:
Revert "Rm all"

This reverts commit 8d2482d840f4d5fa606fbcf3a295a0035c68795f.

dorlneylon commented on internal/core/hash.go:
```diff
@@ -1,12 +1,28 @@
+	h.Write([]byte(s))
+	sum := h.Sum(nil)
+	return uint32(binary.BigEndian.Uint32(sum[:4]))
 }
```
Too complex hashing function ruining performance. Let's consider using siphash: https://en.wikipedia.org/wiki/SipHash, and changing seeds for siphash over changing hashing functions.

dorlneylon commented on internal/core/filter.go:
```diff
@@ -1,31 +1,56 @@
+)
 
 type BloomFilter struct {
 	bs *util.Bitset
-	H  []func(string) int32
+	H  []func(string) uint32
 }
```
Can we instead have a generic parameter for BloomFilter, thus becoming able to accept any hashable type?

jexxor replied:
I'll keep user interface the same. Is it ok as we focus on the BloomFilter functionality?

dorlneylon replied:
It would've been ok as long as it required a lot of effort putting [2 more lines of code](https://github.com/goccy/go-json/) with marshalling.

jexxor replied:
Added `any` data insertion and wrapper with generic (see [here](https://github.com/Jlexender/algo-filter/blob/main/internal/wrapper/filter.go))

dorlneylon commented on internal/core/filter.go:
```diff
@@ -1,31 +1,56 @@
-func NewBloomFilter(size int32) *BloomFilter {
+func NewBloomFilter(size uint32) *BloomFilter {
 	return &BloomFilter{
 		bs: util.NewBitset(size),
-		H: HashList(),
+		H:  HashList(),
```
Could you treat `HashList` as a [Strategy](https://en.wikipedia.org/wiki/Strategy_pattern) programming pattern?

jexxor replied:
It already seems to follow Strategy pattern. If we start to decompose the system, we see that HashList is the list of objects with the same interface, and thus:
![image](https://github.com/user-attachments/assets/22b7ad29-525d-4c44-bd09-fe13768fd1db)

This is similar to Wikipedia's UML. The context is hash calculation inside the filter itself (see [filter.go](https://github.com/Jlexender/algo-filter/blob/main/internal/core/filter.go))
![image](https://github.com/user-attachments/assets/4a47ca56-e715-4b91-a091-2405a0d3acca)

dorlneylon replied:
However, they don't follow a single interface declaring the set of corresponding parameters, i.e. we could put a non-hashing function there. Doesn't sound like good design to me.

jexxor replied:
Though in this particular case function signature `func(string) uint32` can be interpreted as an interface, and by describing this signature we know what is common between these functions. 
 Anyway, I'll more likely switch to Go interfaces when changing hash functionality.

dorlneylon commented on internal/core/filter.go:
```diff
@@ -1,31 +1,56 @@
+	return bf.bs.Size()
+}
+
+func (bf *BloomFilter) List() *util.Bitset {
+	return bf.bs
+}
```
What for?

jexxor replied:
Removed

dorlneylon commented on internal/core/filter_test.go:
```diff
@@ -0,0 +1,88 @@
+	if filter.Exists(wrongString) {
+		t.Logf("expected '%s' to not exist in the filter (false positive?)", wrongString)
+	}
+}
```
Current impl seems to be collision irresistant. Consider using [appropriate](https://en.wikipedia.org/wiki/Bloom_filter#Probability_of_false_positives) number of hashing functions.

jexxor replied:
Added collision resistance afaik (see [here](https://github.com/Jlexender/algo-filter/blob/main/internal/core/hashlist.go))

jexxor committed:
Major update
- collision resistance
- using Go interfaces to perform Strategy pattern
- seeded SipHash for hashing
- removed UI

jexxor committed:
fix size increase on duplicate, added unit tests for filter, cornercases in filter

dorlneylon commented on internal/core/filter.go:
```diff
@@ -1,31 +1,62 @@
-	H  []func(string) int32
+	bs           *util.Bitset
+	hashes       []Hash
+	elementCount uint32
```
```suggestion
	elements uint32
```

dorlneylon replied:
nit, doesn't matter

jexxor committed:
added generic wrapper

dorlneylon commented on internal/core/hashlist.go:
```diff
@@ -0,0 +1,30 @@
+	return list
+}
+
+// hashes := filterSize/elements * ln(2)
+// see https://en.wikipedia.org/wiki/Bloom_filter#Optimal_number_of_hash_functions
+func UpdateList(list []Hash, fsize, elts uint32) []Hash {
+	listLen := xLn2(fsize) / elts
+	if listLen < uint32(len(list)) {
+		list = list[:listLen]
+	}
+	return list
+}
```
Remove, please. `k` is fixed on bloom filter initialization.

Jlexender replied:
why? this operation is efficient and leverages the collision probability (see [here](https://en.wikipedia.org/wiki/Bloom_filter#Optimal_number_of_hash_functions))

dorlneylon commented on internal/core/filter.go:
```diff
@@ -1,31 +1,62 @@
+	for _, hash := range bf.hashes {
+		hashsum := hash.Compute(mapToBytes(data))
+		bitset.Set(hashsum % bitset.Size())
+		bf.hashes = UpdateList(bf.hashes, bf.Size(), bf.elementCount)
```
rm, see further

dorlneylon reviewed (approved):
lgtm, few notes but ok

jexxor committed:
Update README.md

jexxor committed:
Update internal/core/filter.go

Co-authored-by: sodamntired <dorlneylon@gmail.com>

jexxor committed:
fix namings

jexxor committed:
Merge branch 'pr'
